corgidrp/klip_fm.py

```diff
@@ -0,0 +1,462 @@
+    """Injects a fake psf from the CT calibration object into a corgidrp Image with 
+    the desired position and amplitude. 
+
+    Args:
+        frame_in (corgidrp.data.Image): 2D image to inject a fake signal into.
+        ct_calibration (corgidrp.data.CoreThroughputCalibration): CT calibration object containing PSF samples.
+        amp (float): peak pixel amplitude of psf to inject.
+        sep_pix (float): separation from star in pixels to inject 
+        pa_deg (float): position angle to inject (counterclockise from north/up)
+
+    Returns: 
+        corgidrp.data.Image: a copy of the input Image but with a fake PSF injected.
+    """
+
+    frame = frame_in.copy()
+
+    # Get closest psf model
+    frame_roll = frame.ext_hdr['ROLL']
+    rel_pa = pa_deg - frame_roll
+    dx,dy = seppa2dxdy(sep_pix,rel_pa)
+
+    psf_model = get_closest_psf(ct_calibration,
+                                frame.ext_hdr['STARLOCX'],
+                                frame.ext_hdr['STARLOCY'],
+                                dx,dy).copy() 
+
+    # Scale counts
+    peak_count = np.nanmax(psf_model)
+    psf_model *= amp / peak_count
+
+    # Assume PSF is centered in the data cutout for now
+    shape_arr = np.array(psf_model.shape)
+    psf_cenyx_ind = (np.array(shape_arr)/2 - 0.5).astype(int) 
+    psf_cenyx_inframe = np.array([frame.ext_hdr['STARLOCY'],frame.ext_hdr['STARLOCX']]) + np.array([dy,dx])
+    injected_psf_cenyx_ind = np.round(psf_cenyx_inframe).astype(int)
+    starty, startx = injected_psf_cenyx_ind - psf_cenyx_ind
+
+    # Insert into correct frame size array 
+    psf_only_frame = np.zeros_like(frame.data)
+    psf_only_frame[starty:starty+shape_arr[0],startx:startx+shape_arr[1]] = psf_model
+    
+    # # TODO: Calculate subpixel shift:
+    # psf_shift = (0.,0.) # Hardcode 0 shift for now
+    # shifted_psf_only_frame = shift(psf_only_frame,psf_shift)
+
+    # Add to input frame
+    frame.data += psf_only_frame
+
+    psf_cenxy = [psf_cenyx_inframe[1],psf_cenyx_inframe[0]]
+    return frame, psf_model, psf_cenxy
+
+
+def measure_noise(frame, seps_pix, hw, klmode_index=None):
+    """Calculates the noise (standard deviation of counts) of an 
+        annulus at a given separation from the mask center.
+    
+    Args:
+        frame (corgidrp.Image): Image containing data as well as "MASKLOCX/Y" in header
+        seps_pix (np.array of float): Separations (in pixels from mask center) at which to calculate 
+            the noise level.
+        klmode_index (int, optional): If provided, returns only the noise values for the KL mode with 
+            the given index. I.e. klmode_index=0 would return only the values for the first KL mode 
+            truncation choice.  If None (by default), all indices are returned.
+        hw (float): halfwidth of the annulus to use for noise calculation.
+
+    Returns: np.array 
+    """
+
+    cenx, ceny = (frame.ext_hdr['MASKLOCX'],frame.ext_hdr['MASKLOCY'])
```

Comment:
I think technically we want to compute noise centered around the star, not the mask (so STARLOCX/Y)